src/deploy/functions/params.ts

```diff
@@ -0,0 +1,342 @@
+ */
+function resolveDefaultCEL(
+  type: string,
+  expr: CEL,
+  currentEnv: Record<string, ParamValue>
+): ParamValue {
+  const deps = dependenciesCEL(expr);
+  const allDepsFound = deps.every((dep) => {
+    return Object.keys(currentEnv).includes(dep);
+  });
+  if (!allDepsFound) {
+    throw new FirebaseError(
+      "Build specified parameter with un-resolvable default value " +
+        expr +
+        "; dependencies missing."
+    );
+  }
+
+  switch (type) {
+    case "string":
+      return resolveString(expr, currentEnv);
+    case "int":
+      return resolveInt(expr, currentEnv);
+    default:
+      throw new FirebaseError(
+        "Build specified parameter with default " + expr + " of unsupported type"
+      );
+  }
+}
+/**
+ * Tests whether a mooted ParamValue literal is of the correct type to be the value for a Param.
+ */
+function canSatisfyParam(param: Param, value: ParamValue): boolean {
+  if (param.type === "string") {
+    return typeof value === "string";
+  } else if (param.type === "int") {
+    return typeof value === "number" && Number.isInteger(value);
+  } else {
+    assertExhaustive(param);
+  }
+}
+
+/**
+ * A param defined by the SDK may resolve to:
+ * - the value of a Cloud Secret in the same project with name == param name (not implemented yet), but only if it's a SecretParam
+ * - a literal value of the same type already defined in one of the .env files with key == param name
+ * - the value returned by interactively prompting the user
+ *   - the default value of the prompt comes from the SDK via param.default, which may be a literal value or a CEL expression
+ *   - if the default CEL expression is not resolvable--it depends on a param whose value is not yet known--we throw an error
+ *   - yes, this means that the same set of params may or may not throw depending on the order the SDK provides them to us in
+ *   - after prompting, the resolved value of the param is written to the most specific .env file available
+ */
+export async function resolveParams(
+  params: Param[],
+  projectId: string,
+  userEnvs: Record<string, ParamValue>
+): Promise<Record<string, ParamValue>> {
+  const paramValues: Record<string, ParamValue> = {};
+
+  for (const param of params) {
+    if (userEnvs.hasOwnProperty(param.param)) {
+      if (canSatisfyParam(param, userEnvs[param.param])) {
+        paramValues[param.param] = userEnvs[param.param];
+        continue;
+      } else {
+        throw new FirebaseError(
+          "Parameter " +
+            param.param +
+            " resolved to value from dotenv files " +
+            userEnvs[param.param] +
+            " of wrong type"
+        );
+      }
+    }
+
+    let paramDefault: ParamValue | undefined = param.default;
+    if (paramDefault && isCEL(paramDefault)) {
+      paramDefault = resolveDefaultCEL(param.type || "string", paramDefault, paramValues);
```

Comment:
Why did you need the `|| "string"`? The type of `param.type` is `"string" | "int"`. Either type should be optional in the definition of `Param` or you shouldn't have a backup and this will fail to compile if you ever _did_ need a fallback value.

Reply:
I think this might have just been when it was `type?` instead of 'type' and I threw that in to make the compiler happy without realizing what I was doing. Fixed, now.